File: android/app/src/main/java/info/bagen/rust/plaoc/microService/ipc/NativeIpc.kt
package info.bagen.rust.plaoc.microService.ipc

import info.bagen.rust.plaoc.microService.helper.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch

inline fun debugNativeIpc(tag: String, msg: Any = "", err: Throwable? = null) =
    printdebugln("native-ipc", tag, msg, err)

class NativeIpc(
    val port: NativePort<IpcMessage, IpcMessage>,
    override val remote: MicroModuleInfo,
    private val role_type: IPC_ROLE,
) : Ipc() {
    override val role get() = role_type.role
    override fun toString(): String {
        return super.toString() + "@NativeIpc"
    }

    override val supportRaw = true
    override val supportBinary = true

    init {
        port.onMessage { message ->
            _messageSignal.emit(IpcMessageArgs(message, this@NativeIpc))
        }
        GlobalScope.launch(ioAsyncExceptionHandler) {
            port.start()
        }
    }


    override suspend fun _doPostMessage(data: IpcMessage) {
        port.postMessage(data)
    }

    override suspend fun _doClose() {
        port.close()
    }
}

class NativePort<I, O>(
    private val channel_in: Channel<I>,
    private val channel_out: Channel<O>,
    private val closePo: PromiseOut<Unit>
) {
    companion object {
        private var uid_acc = 1;
    }

    private val uid = uid_acc++
    override fun toString() = "#p$uid"

    private var started = false
    suspend fun start() {
        if (started || closePo.isFinished) return else started = true

        debugNativeIpc("port-message-start/$this")
        for (message in channel_in) {
            debugNativeIpc("port-message-in/$this << $message")
            _messageSignal.emit(message)
            debugNativeIpc("port-message-waiting/$this")
        }
        debugNativeIpc("port-message-end/$this")
    }


    private val _closeSignal = SimpleSignal()

    fun onClose(cb: SimpleCallback) = _closeSignal.listen(cb)

    fun close() {
        if (!closePo.isFinished) {
            closePo.resolve(Unit)
            debugNativeIpc("port-closing/$this")
        }
    }

    /**
     * ?????? close ??????????????????????????????????????????????????????
     */
    init {
        GlobalScope.launch(ioAsyncExceptionHandler) {
            closePo.waitPromise()
            channel_out.close()
            _closeSignal.emit()
            debugNativeIpc("port-closed/${this@NativePort}")
        }

    }


    private val _messageSignal = Signal<I>()

    /**
     * ????????????????????????????????????
     */
    suspend fun postMessage(msg: O) {
        debugNativeIpc("message-out/$this >> $msg")
        channel_out.send(msg)
    }

    /**
     * ????????????
     */
    fun onMessage(cb: Callback<I>) = _messageSignal.listen(cb)
}

class NativeMessageChannel<T1, T2> {
    /**
     * ????????????????????????????????????????????????????????????
     */
    private val closePo = PromiseOut<Unit>()
    private val channel1 = Channel<T1>()
    private val channel2 = Channel<T2>()
    val port1 = NativePort(channel1, channel2, closePo)
    val port2 = NativePort(channel2, channel1, closePo)
}

File: ios/BFExplorer/Plaoc/bfs/ipc/NativeIpc.swift
//
//  NativeIpc.swift
//  BFExplorer
//
//  Created by ui08 on 2023/2/20.
//

import Foundation
import Combine

class NativeIpc: Ipc {
    let port: NativePort<IpcMessage, IpcMessage>
    
    init(port: NativePort<IpcMessage, IpcMessage>, remote: MicroModule, role: IPC_ROLE) {
        self.port = port
        super.init()
        self.remote = remote
        self.role = role.rawValue
        self.support_raw = true
        self.support_bianry = true
        
        _ = port.onMessage { message in
            await self._messageSignal.emit((message, self))
            return nil
        }
        
        port.start()
    }
    
    override func _doPostMessage(data: IpcMessage) async {
        await port.postMessage(msg: data)
    }
    
    override func _doClose() async {
        port.close()
    }
    
    override func toString() -> String {
        super.toString() + "@NativeIpc"
    }
}

class NativePortUid {
    static var uid_acc = 1
}
class NativePort<I, O> {
    private let channel_in: PassthroughSubject<I, Never>
    private let channel_out: PassthroughSubject<O, Never>
    private let closePo: PromiseOut<()>
    private var cancellable: AnyCancellable?
    
    private var uid = NativePortUid.uid_acc++
    
    func toString() -> String {
        "#p\(uid)"
    }
    
    init(channel_in: PassthroughSubject<I, Never>, channel_out: PassthroughSubject<O, Never>, closePo: PromiseOut<()>) {
        self.channel_in = channel_in
        self.channel_out = channel_out
        self.closePo = closePo
        
        Task {
            await closePo.waitPromise()
            closing = true
            cancellable?.cancel()
            await _closeSignal.emit(())
        }
    }
    
    private var started = false
    
    func start() {
        if started || closing {
            return
        } else {
            started = true
        }
        
        print("port-message-start/\(self)")
        cancellable = channel_in.sink(receiveValue: { message in
            Task {
                print("port-message-in/\(self) << \(message)")
                await self._messageSignal.emit(message)
                print("port-message-waiting/\(self)")
            }
        })
        print("port-message-end/\(self)")
    }
    
    private let _closeSignal = Signal<()>()
    
    private var closing = false
    
    func close() {
        if !closePo.isFinished {
            closePo.resolve(())
            print("port-closing/\(self)")
        }
    }
    
    private let _messageSignal = Signal<I>()
    
    func postMessage(msg: O) async {
        channel_out.send(msg)
        
        if let msg = msg as? IpcMessage {
            print(msg.type)
        }
    }
    
    func onMessage(cb: @escaping (I) async -> SIGNAL_CTOR?) -> () async -> Bool {
        self._messageSignal.listen(cb)
    }
}


struct NativeMessageChannel<T1, T2> {
    private let closePo = PromiseOut<()>()
    private let channel1 = PassthroughSubject<T1, Never>()
    private let channel2 = PassthroughSubject<T2, Never>()
    let port1: NativePort<T1, T2>
    let port2: NativePort<T2, T1>
    
    init() {
        port1 = NativePort(channel_in: channel1, channel_out: channel2, closePo: closePo)
        port2 = NativePort(channel_in: channel2, channel_out: channel1, closePo: closePo)
    }
}
